File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/KeyfileExportViewModel.kt
// "Back up your account key" — reads the UMK out of the Keystore and
// wraps it into a passphrase-encrypted `.flagshipkey` file, byte-
// compatible with packages/protocol/src/keyfile.ts. Mirror of the iOS
// KeyfileExportViewModel.
//
// The view holds the three required acknowledgments + the passphrase;
// this VM only validates strength and runs the wrap. The produced file
// text is handed back so the host can save it via the Storage Access
// Framework / a share intent — we never write it anywhere on our own.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.keystore.Keyfile
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface KeyfileExportPhase {
    data object Idle : KeyfileExportPhase
    data object Working : KeyfileExportPhase
    /** The keyfile text is ready; the host saves it (SAF / share). */
    data class Ready(val text: String) : KeyfileExportPhase
    data class Failed(val message: String) : KeyfileExportPhase
}

class KeyfileExportViewModel(
    val username: String,
    val accountId: String? = null,
    /** Seam so tests can supply the UMK without piercing the Keystore.
     *  Defaults to the real (active-profile) UMK seed read. */
    private val readUmkSeed: () -> ByteArray = { com.flagshipserver.app.keystore.Keystore.currentUmkSeed() },
    /** Injectable clock for deterministic createdAt in tests. */
    private val nowIso: () -> String = { Keyfile.nowIso() },
) : ViewModel() {

    private val _phase = MutableStateFlow<KeyfileExportPhase>(KeyfileExportPhase.Idle)
    val phase: StateFlow<KeyfileExportPhase> = _phase.asStateFlow()

    private val _passphrase = MutableStateFlow("")
    val passphrase: StateFlow<String> = _passphrase.asStateFlow()
    fun setPassphrase(v: String) { _passphrase.value = v }

    private val _confirm = MutableStateFlow("")
    val confirmPassphrase: StateFlow<String> = _confirm.asStateFlow()
    fun setConfirmPassphrase(v: String) { _confirm.value = v }

    // The three required acknowledgments. All must be true before
    // "Create backup file" enables.
    private val _ackControl = MutableStateFlow(false)
    val ackControl: StateFlow<Boolean> = _ackControl.asStateFlow()
    fun setAckControl(v: Boolean) { _ackControl.value = v }

    private val _ackOffline = MutableStateFlow(false)
    val ackOffline: StateFlow<Boolean> = _ackOffline.asStateFlow()
    fun setAckOffline(v: Boolean) { _ackOffline.value = v }

    private val _ackNoRecovery = MutableStateFlow(false)
    val ackNoRecovery: StateFlow<Boolean> = _ackNoRecovery.asStateFlow()
    fun setAckNoRecovery(v: Boolean) { _ackNoRecovery.value = v }

    val acknowledged: Boolean
        get() = _ackControl.value && _ackOffline.value && _ackNoRecovery.value

    val passphraseStrong: Boolean
        get() = isStrong(_passphrase.value)

    val passphrasesMatch: Boolean
        get() = _confirm.value.isNotEmpty() && _passphrase.value == _confirm.value

    /** The "Create backup file" CTA enables only when everything lines
     *  up: strong passphrase, confirmation matches, all three acks. */
    val canCreate: Boolean
        get() = passphraseStrong && passphrasesMatch && acknowledged

    /** Suggested filename for the save dialog: `<username>.flagshipkey`. */
    val suggestedFilename: String
        get() = (username.ifEmpty { "account" }) + ".flagshipkey"

    /** Read the UMK + wrap it. On success → Ready(text). */
    fun createBackup() {
        if (!canCreate) return
        _phase.value = KeyfileExportPhase.Working
        try {
            val seed = readUmkSeed()
            val meta = Keyfile.Meta(
                username = username,
                accountId = accountId,
                createdAt = nowIso(),
            )
            val text = Keyfile.wrap(umkSeed = seed, passphrase = _passphrase.value, meta = meta)
            _phase.value = KeyfileExportPhase.Ready(text)
        } catch (t: Throwable) {
            _phase.value = KeyfileExportPhase.Failed("Couldn't create the backup file: ${t.message}")
        }
    }

    fun reset() {
        _phase.value = KeyfileExportPhase.Idle
    }

    companion object {
        /** A simple, defensible strength rule. The keyfile floor is 8;
         *  we ask for more in the UI since this file is the keys to the
         *  whole account: >= 12 chars and at least 3 of 4 character
         *  classes (lower / upper / digit / symbol). Mirror of the iOS
         *  KeyfileExportViewModel.isStrong. */
        fun isStrong(s: String): Boolean {
            if (s.length < 12) return false
            val lower = s.any { it.isLowerCase() }
            val upper = s.any { it.isUpperCase() }
            val digit = s.any { it.isDigit() }
            val symbol = s.any { !it.isLetterOrDigit() && !it.isWhitespace() }
            val classes = listOf(lower, upper, digit, symbol).count { it }
            return classes >= 3
        }
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/KeyfileExportViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipCore

/// Drives "Back up your account key" — reads the UMK out of the
/// Keystore (behind the biometric gate) and wraps it into a
/// passphrase-encrypted `.flagshipkey` file, byte-compatible with
/// `packages/protocol/src/keyfile.ts`.
///
/// The view holds the required control acknowledgment + the passphrase;
/// this VM only validates strength and runs the wrap. The produced file
/// text is handed back so the host can present a share sheet — we never
/// write it anywhere on our own.
@MainActor
@Observable
public final class KeyfileExportViewModel {
    public enum Phase: Equatable, Sendable {
        case idle
        case working
        /// The keyfile text is ready; host presents the share sheet.
        case ready(text: String)
        case failed(String)
    }

    public private(set) var phase: Phase = .idle
    public var passphrase: String = ""
    public var confirmPassphrase: String = ""

    /// Required before the "Create backup file" button enables.
    public var ackControl: Bool = false

    public let username: String
    public let accountId: String?

    /// Seam so tests can supply the UMK without piercing the Secure
    /// Enclave. Defaults to the real Keystore read (fires biometrics).
    private let readUMK: @MainActor (String) async throws -> SymmetricKey

    public init(
        username: String,
        accountId: String? = nil,
        readUMK: @escaping @MainActor (String) async throws -> SymmetricKey = { reason in
            try await Keystore.currentUMK(reason: reason)
        }
    ) {
        self.username = username
        self.accountId = accountId
        self.readUMK = readUMK
    }

    /// The user acknowledged that the file + passphrase control the account.
    public var acknowledged: Bool {
        ackControl
    }

    /// The backup UI requires at least 12 characters.
    public var passphraseStrong: Bool {
        Self.isStrong(passphrase)
    }

    public var passphrasesMatch: Bool {
        !confirmPassphrase.isEmpty && passphrase == confirmPassphrase
    }

    /// The "Create backup file" CTA is enabled only when everything
    /// lines up: minimum-length passphrase, confirmation matches, and the
    /// acknowledgment is checked.
    public var canCreate: Bool {
        passphraseStrong && passphrasesMatch && acknowledged
    }

    /// A simple minimum-length rule shared by the UI and action layer.
    static func isStrong(_ s: String) -> Bool {
        s.count >= 12
    }

    /// Suggested filename for the share sheet: `<username>.flagshipkey`.
    public var suggestedFilename: String {
        let safe = username.isEmpty ? "account" : username
        return "\(safe).flagshipkey"
    }

    /// Read the UMK + wrap it. On success → `.ready(text:)`.
    public func createBackup() async {
        guard canCreate else { return }
        phase = .working
        do {
            let umk = try await readUMK("Back up your Flagship account key")
            let meta = Keyfile.Meta(
                username: username,
                accountId: accountId,
                createdAt: ISO8601DateFormatter.flagshipUTC.string(from: Date())
            )
            let text = try Keyfile.wrap(umkSeed: umk, passphrase: passphrase, meta: meta)
            phase = .ready(text: text)
        } catch {
            phase = .failed("Couldn't create the backup file: \(error.localizedDescription)")
        }
    }

    public func reset() {
        phase = .idle
    }
}

extension ISO8601DateFormatter {
    /// UTC ISO-8601 with milliseconds — matches the `createdAt` shape in
    /// the golden keyfile (`2026-05-25T00:00:00.000Z`).
    static let flagshipUTC: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        f.timeZone = TimeZone(identifier: "UTC")
        return f
    }()
}
